README
NAME
    Mojolicious::Plugin::Authorization - A plugin to make Authorization a
    bit easier

VERSION
    version 1.03

SYNOPSIS
        use Mojolicious::Plugin::Authorization
        $self->plugin('Authorization' => {
            'has_priv'   => sub { ... },
            'is_role'    => sub { ... },
            'user_privs' => sub { ... },
            'user_role'  => sub { ... },
        });
        if ($self->has_priv('delete_all', { optional => 'extra data stuff' })) {
            ...
        }

DESCRIPTION
    A very simple API implementation of role-based access control (RBAC).
    This plugin is only an API you will have to do all the work of setting
    up your roles and privileges and then provide four subs that are used by
    the plugin. The plugin expects that the current session will be used to
    get the role its privileges. It also assumes that you have already been
    authenticated and your role set. That is about it you are free to
    implement any system you like.

METHODS
  has_priv('privilege', $extra_data) or has_privilege('privilege', $extra_data)
    'has_priv' and 'has_privilege' will use the supplied `has_priv'
    subroutine ref to check if the current session has the given privilege.
    Returns true when the session has the privilege or false otherwise. You
    can pass additional data along in the extra_data hashref and it will be
    passed to your `has_priv' subroutine as-is.

  is('role',$extra_data) / is_role('role,$extra_data)
    'is' / 'is_role' will use the supplied `is_role' subroutine ref to check
    if the current session is the given role. Returns true when the session
    has privilege or false otherwise. You can pass additional data along in
    the extra_data hashref and it will be passed to your `is_role'
    subroutine as-is.

  privileges($extra_data)
    'privileges' will use the supplied `user_privs' subroutine ref and
    return the privileges of the current session. You can pass additional
    data along in the extra_data hashref and it will be passed to your
    `user_privs' subroutine as-is. The returned data is dependant on the
    supplied `user_privs' subroutine.

  role($extra_data)
    'role' will use the supplied `user_role' subroutine ref and return the
    role of the current session. You can pass additional data along in the
    extra_data hashref and it will be passed to your `user_role' subroutine
    as-is. The returned data is dependant on the supplied `user_role'
    subroutine.

CONFIGURATION
    The following options must be set for the plugin:

    has_priv (REQUIRED) A coderef for checking to see if the current session
    has a privilege (see "HAS PRIV").
    is_role (REQUIRED) A coderef for checking to see if the current session
    is a certain role (see "IS ROLE").
    user_privs (REQUIRED) A coderef for returning the privileges of the
    current session (see "PRIVILEGES").
    user_role (REQUIRED) A coderef for retiring the role of the current
    session (see "ROLE").

  HAS PRIV
    'has_priv' is used when you need to confirm that the current session has
    the given privilege. The coderef you pass to the `has_priv'
    configuration key has the following signature:

        sub {
            my ($app, $privilege,$extradata) = @_;
            ...
        }

    You must return either 0 for a fail and 1 for a pass. This allows
    `ROUTING VIA CONDITION' to work correctly.

  IS / IS ROLE
    'is' / 'is_role' is used when you need to confirm that the current
    session is set to the given role. The coderef you pass to the `is_role'
    configuration key has the following signature:

        sub {
            my ($app, $role, $extradata) = @_;
            ...
            return $role;
        }

    You must return either 0 for a fail and 1 for a pass. This allows
    `ROUTING VIA CONDITION' to work correctly.

  PRIVILEGES
    'privileges' is used when you need to get all the privileges of the
    current session. The coderef you pass to the `user_privs' configuration
    key has the following signature:

        sub {
            my ($app,$extradata) = @_;
            ...
            return $privileges;
        }

    You can return anything you want. It would normally be an arrayref of
    privileges but you are free to return a scalar, hashref, arrayref,
    blessed object, or undef.

  ROLE
    'role' is used when you need to get the role of the current session. The
    coderef you pass to the `user_privs' configuration key has the following
    signature:

        sub {
            my ($app,$extradata) = @_;
            ...
            return $role;
        }
    
    You can return anything you want. It would normally be just a scalar but
    you are free to return a scalar, hashref, arrayref, blessed object, or
    undef.

EXAMPLES
    For a code example using this, see the t/01-functional.t test, it uses
    Mojolicious::Lite and this plugin.

ROUTING VIA CONDITION
    This plugin also exports a routing condition you can use in order to
    limit access to certain documents to only sessions that have a
    privilege.

        $r->route('/delete_all')->over(has_priv => 'delete_all')->to('mycontroller#delete_all');
        my $delete_all_only = $r->route('/members')->over(has_priv => 'delete_all')->to('members#delete_all');
        $delete_all_only->route('delete')->to('members#delete_all');

    If the session does not have the 'delete_all' privilege, these routes
    will not be considered by the dispatcher and unless you have set up a
    catch-all route, a 404 Not Found will be generated instead.

    Another condition you can use to limit access to certain documents to
    only those sessions that have a role.

        $r->route('/view_all')->over(is => 'ADMIN')->to('mycontroller#view_all');
        my $view_all_only = $r->route('/members')->over(is => 'view_all')->to('members#view_all');
        $view_all_only->route('view')->to('members#view_all');

    If the session is not the 'ADMIN' role, these routes will not be
    considered by the dispatcher and unless you have set up a catch-all
    route, a 404 Not Found will be generated instead. This behavior is
    similar to the "has" condition.

ROUTING VIA CALLBACK
    It is not recommended to route un-authorized requests to anything but a
    404 page. If you do route to some sort of 'You are not allowed page' you
    are telling a hacker that the URL was correct while the 404 tells them
    nothing. This is just my opinion.

SEE ALSO
    Mojolicious::Sessions, Mojocast 3: Authorization

AUTHOR
    John Scoles, `<byterock at hotmail.com>'

BUGS / CONTRIBUTING
    Please report any bugs or feature requests through the web interface at
    https://github.com/byterock/mojolicious-plugin-authorization/issues.

SUPPORT
    You can find documentation for this module with the perldoc command.
    perldoc Mojolicious::Plugin::Authorization You can also look for
    information at:

    * AnnoCPAN: Annotated CPAN documentation
    http://annocpan.org/dist/Mojolicious-Plugin-Authorization
    * CPAN Ratings
    http://cpanratings.perl.org/d/Mojolicious-Plugin-Authorization
    * Search CPAN
    http://search.cpan.org/dist/Mojolicious-Plugin-Authorization/

ACKNOWLEDGEMENTS
    Ben van Staveren (madcat)

        -   For 'Mojolicious::Plugin::Authentication' which I used as a guide in writing up this one.
    
    Chuck Finley

        -   For staring me off on this.
    
    Abhijit Menon-Sen

        -   For the routing suggestions
    
    Roland Lammel

        -   For some other good suggestions

LICENSE AND COPYRIGHT
    Copyright 2012 John Scoles. This program is free software; you can
    redistribute it and/or modify it under the terms of either: the GNU
    General Public License as published by the Free Software Foundation; or
    the Artistic License. See http://dev.perl.org/licenses/ for more
    information.



Changes
Revision history for Mojolicious-Plugin-Authorization

Changes in 1.01
    Initail version
Changes in 1.02
    Fixed the pod
Changes in 1.03
    3'rd Times a charm
Changes in 1.03.01
    Fix up on some the the tests. No no fall though errors
Changes in 1.03.02
    Fix up on some the the tests. Hash keys now sorted from Renee https://github.com/reneeb
Changes in 1.04
    README Changes from leejo https://github.com/leejo
    Add is_role helper and condion as well as test for same from leejo https://github.com/leejo
    Add fail_render option on 404 and test from leejo https://github.com/leejo
    Typo fixes
Changes in 1.05
    Typo fixes
    Fix for internal calling profile of 'privileges' and 'role'
    Added extra data via conditional routing by Matt Altus committed on 19 Mar 
    Remove Build.PL from MANIFEST not needed


Build.PL
use strict;
use warnings;
use Module::Build 0.38;
my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.38",
    "Mojolicious::Lite" => 0,
    "Test::Mojo" => 0,
    "Test::More" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "configure_requires" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Module::Build" => "0.38"
  },
  "dist_abstract" => "A plugin to make authorization a bit easier",
  "dist_author" => [
    "John Scoles <byterock\@hotmail.com>"
  ],
  "dist_name" => "Mojolicious-Plugin-Authorization",
  "dist_version" => "1.05",
	"license" => "perl",
  "module_name" => "Mojolicious::Plugin::Authorization",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Mojo::Base" => 0
  },
  "script_files" => [],
  "meta_merge" => {
    "resources" => {
      "repository" => 'https://github.com/byterock/mojolicious-plugin-authorization'
    }
  },
);
my $build = Module::Build->new(%module_build_args);
$build->create_build_script;


Makefile.PL
use strict;
use warnings;
use ExtUtils::MakeMaker 6.30;
my %WriteMakefileArgs = (
  "ABSTRACT" => "A plugin to make authorization a bit easier",
  "AUTHOR" => "John Scoles <byterock\@hotmail.com>",
  "BUILD_REQUIRES" => {
    "Module::Build" => "0.38",
    "Mojolicious::Lite" => 0,
    "Test::Mojo" => 0,
    "Test::More" => 0,
    "Test::Exception" => 0.32,
    "Test::Deep"      => 0.113,
    "strict" => 0,
    "warnings" => 0
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Module::Build" => "0.38"
  },
  "DISTNAME" => "Mojolicious-Plugin-Authorization",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Mojolicious::Plugin::Authorization",
  "PREREQ_PM" => {
    "Mojo::Base" => 0
  },
  "VERSION" => "1.05",
  "test" => {
    "TESTS" => "t/*.t"
  },
  "META_MERGE" => {
    "resources" => {
      "repository" => 'https://github.com/byterock/mojolicious-plugin-authorization',
      "bugtracker" => 'https://github.com/byterock/mojolicious-plugin-authorization/issues',
      "homepage"   => 'https://metacpan.org/module/Mojolicious::Plugin::Authorization',
    },
  },
);
unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}
delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };
WriteMakefile(%WriteMakefileArgs);


example/miniauthorfile.pm
package
   miniauthorfile;     # hide from PAUSE
use strict;
use warnings;
use warnings FATAL => qw{ uninitialized };
use autodie;
use 5.10.0;
################################################################
=pod
=head1 Title
  miniautorfile.pm --- mini data base for a role-based access control (RBAC) file.
=head1 Invocation
  $ perl miniautorfile.pm
shows off how this module works.
=head1 Versions
  0.0: April 11 2012
=cut
################################################################
# file format: role:privilege1:privilege2:privilege3
#              role1:privilege1:privilege3
################################################################
sub new {
  my $class = shift;
  my ($authorfile)= @_;
  (-e $authorfile) or die "You must create a user-readable and user-writable authorization file first.\n";
  ## load persistent role information from an existing authorization file
  my %roles;
  open(my $FIN, "<", $authorfile);
  while (<$FIN>) {
    (/^\#/) and next; ## skip comments
    (/\w/) or next;  ## skip empty lines
    (!/([\w :\\])/) and die "Your authorization file has a non-word character ($1), other than : and \\ on line $.: $_\n";
    my @values= split(/:/);
    my $role = shift(@values);
    my $privs;
    foreach my $priv (@values){
       $privs->{$priv} = 1;
    }
    $roles{$role}= $privs;
  }
  close($FIN);
  return bless({ authorfile => $authorfile, %roles }, $class);
}
################################################################
sub set_role {
  my $self = shift;
  my ($session,$role) = @_;
  #$session->{'role_privs'} = $roles{$role};
  $session->{'role'}=$role;
#  return ((exists($self->{$_[0]}))?($_[0]):undef);
}
1;


example/showoff-authorization.pl
#!/usr/bin/env perl
use strict;
use warnings;
use warnings FATAL => qw{ uninitialized };
use autodie;
# Disable IPv6, epoll and kqueue
BEGIN { $ENV{MOJO_NO_IPV6} = $ENV{MOJO_POLL} = 1 }
use Mojolicious::Lite;
=pod
=head1 Title
  showoff-authorization.pl --- an example of the Mojolicious::Plugin::Authorization module by John Scoles
=head1 Invocation
  $ perl showoff-authorization.pl daemon
=head1 Notes
My first crack at a Mojo plugin a steal from Ben van Staveren's Authentication so I owe him and some others
a great note of thanks
Like Authentication this is a very a simple application. It supplies the framwork and you have to give it
the guts which this little progam shows.
I did not add in any Authentication as that is up to you to build. In this test I just assume you are
authenticated on the session ans that session has a role hash on it.
=head1 Versions
  0.1: May 01 2012
=cut
################################################################
### miniauthorfile.pm lays out basic functionality for the miniauthorfile
use miniauthorfile;
my $roles = miniauthorfile->new('miniauthorfile.txt');
################################################################
plugin 'authorization', {

  has_priv => sub {
     my $self = shift;
     my ($priv, $extradata) = @_;
     return 0
       unless($self->session('role'));
     my $role  = $self->session('role');
     my $privs = $roles->{$role};
     return 1
       if exists($privs->{$priv});
     return 0;
  },
  is_role => sub {
    my $self = shift;
    my ($role, $extradata) = @_;
    return 0
       unless($self->session('role'));
    return 1
       if ($self->session('role') eq $role);
   return 0;
  },
  user_privs => sub {
    my $self = shift;
    my ($extradata) = @_;
    return []
       unless($self->session('role'));
    my $role  = $self->session('role');
    my $privs = $roles->{$role};
    return keys(%{$privs});
  },
  user_role => sub {
    my $self = shift;
    my ($extradata) = @_;
    return $self->session('role');
  },
 };
################################################################
get '/' => sub {
  my $self = shift;
  unless($self->session('role')){
    $self->session('role'=>'guest');
  }
  $self->render('index');  ## index needs to be named to match '/'
};
get '/dogshow' => sub {
  my $self = shift;
  unless ($self->has_priv('view')) {
     $self->render('index');
  }
  else{
     $self->stash('role_name'=> $self->role());
     $self->render('dogshow');
  }
};
get '/change/:role' => sub {
  my $self = shift;
  my $role =  $self->param('role');
  $roles->set_role($self->session,$role);
  $self->stash('role_name'=> $self->role());
  $self->render('dogshow');
 # $self->render(template);  ## this is called automatically
};
get '/view' => sub {
  my $self = shift;
  unless ($self->has_priv('view')) {
     $self->render('index');
  }
 # $self->render(template);  ## this is called automatically
};
get '/herd' => sub {
  my $self = shift;
  unless ($self->has_priv('herd')) {
     $self->render('not_allowed');
  }
};
get '/judge' => sub {
  my $self = shift;
  $self->render('not_allowed')
    unless ($self->has_priv('judge'));
  $self->render('all_glory')
    if ($self->is("hypnotoad"))
};
############ these two subs can show you what you can do now, based on authenticated status
get '/my_privs/' => sub {
  my $self = shift;
  $self->render('not_allowed')
    unless ($self->session('role'));
  my @privs = $self->privileges();
  $self->stash('privs'=> \@privs);
};
## /condition/authonly exists as a webpage ONLY after authentication
app->secret('All GLORY to the Hypnotoad!!');  # used for cookies and persistence
app->start();
################################################################
__DATA__
@@ index.html.ep
% layout 'default';
% title 'Root';
<h2> Top Index Page</h2>
<p>The purpose of this little web app is to show an example of <a href="http://mojolicio.us/">Mojolicious</a> and its <a href="http://search.cpan.org/~madcat/Mojolicious-Plugin-Authorization/">Mojolicious::Authorization module</a> by John Scoles.</p>
<p>Go to the trials as a <a href="/change/guest">Guest</a>.</p>
<p>Go to the trials as a <a href="/change/dog">Dog</a>.</p>
<p>Go the trials as a <a href="/change/judge">Judge</a>.</p>
<p>Go the trials as The <a href="/change/hypnotoad">Hypnotoad</a>.</p>
@@ dogshow.html.ep
% layout 'default';
% title 'Pan Galatic Sheep Dog Trials';
<p>Welcome "<%= $role_name %>" to the the Pan Galatic Sheep Dog Trials.</p>
<a href="/">Go home</a><br>
<a href="/view">View a Trial</a><br>
<a href="/herd">Herd some Sheep</a><br>
<a href="/judge">Judge a trial</a><br>
<a href="/my_privs">What are my Privleges</a>
@@ view.html.ep
% layout 'default';
% title 'View Trials';
<h1>Enjoy the Trials</h1>
<p>He's good.</p>
<p>But our real compition is the Hypnotoad</p>
@@ herd.html.ep
% layout 'default';
% title 'Herd Some Sheep';
<h1>Heard Some Sheep</h1>
<p>Woof, Bow-Wow</P>
<p>eye-ball</p>
<p>run~chase</p>
<p>put sheep in pen</p>
@@ judge.html.ep
% layout 'default';
% title 'Judge a Dog';
<h1>Judge a Dog</h1>
<p>5.8</p>
<p>5.9</p>
<p>5.8</p>
<p>5.7</p>
and
<p>4.9</p>
from the Russian Judge
@@ my_privs.html.ep
% layout 'default';
% title 'My Privleges';
<h1>Privleges</h1>
%foreach my $priv (@{$privs}) {
 <%==$priv%> <br>
%}
@@ all_glory.html.ep
% layout 'default';
% title 'Judge a Dog';
<h1>Judge a Dog</h1>
<p>And the winner is </p>
<h1><blink>The Hypnotoad</blink></h1>
<p>All Gloy to the Hypnotoad</p>
Clap-Clap-Clap
@@ not_allowed.html.ep
% layout 'default';
% title 'Page Unavailable';
<h1>I am sorry do to interferance from suicide booths on 'Eminiar VII' you cannot get to this page</h1>
@@ layouts/default.html.ep
<!DOCTYPE html>
<html>
  <head>
    <title><%= title %></title>
  </head>
  <body>
    <hr />
    <h1> Mojolicious: <%= $0 %>: <%= title %> </h1>
    <hr />
    <%= content %>
    <hr />
  </body>
</html>


lib/Mojolicious/Plugin/Authorization.pm
package Mojolicious::Plugin::Authorization;
BEGIN {
  $Mojolicious::Plugin::Authorization::VERSION = '1.05';
}
use Mojo::Base 'Mojolicious::Plugin';
# The dog is good, but our real competition is the Hypnotoad.
sub register {
    my ($self, $app, $args) = @_;
    $args ||= {};
    for my $sub_ref ( qw/ has_priv is_role user_privs user_role / ) {
        die __PACKAGE__, ": missing '$sub_ref' subroutine ref in parameters\n"
            unless $args->{$sub_ref} && ref($args->{$sub_ref}) eq 'CODE';
    }
    my $has_priv_cb    = $args->{has_priv};
    my $is_role_cb     = $args->{is_role};
    my $user_privs_cb  = $args->{user_privs};
    my $user_role_cb   = $args->{user_role};
    my $fail_render    = $args->{fail_render};
    $app->routes->add_condition(has_priv => sub {
        my ($r, $c, $captures, $priv) = @_;
        my $extradata;
        ($priv, $extradata) = @$priv if (ref($priv) eq 'ARRAY');
        my $res = ($priv && $has_priv_cb->($c,$priv,$extradata)) ? 1 : 0;
        $c->render( %$fail_render ) if $fail_render && ! $res;
        return $res;
    });
    for my $helper_name ( qw/ is_role is / ) {
        $app->routes->add_condition($helper_name => sub {
            my ($r, $c, $captures, $role) = @_;
            my $extradata;
            ($role, $extradata) = @$role if (ref($role) eq 'ARRAY');
            my $res = ($role && $is_role_cb->($c,$role,$extradata)) ? 1 : 0;
            $c->render( %$fail_render ) if $fail_render && ! $res;
            return $res;
        });
    }
    $app->helper(privileges => sub {
        my ($c, $extradata) = @_;
        return $user_privs_cb->($c, $extradata);
    });
    for my $helper_name ( qw/ has_priv has_privilege / ) {
        $app->helper($helper_name => sub {
            my ($c, $priv, $extradata) = @_;
            my $has_priv = $has_priv_cb->($c, $priv, $extradata);
            return $has_priv;
        });
    }
    for my $helper_name ( qw/ is is_role / ) {
        $app->helper($helper_name => sub {
            my ($c, $role, $extradata) = @_;
            return $is_role_cb->($c, $role, $extradata);
        });
    }
    $app->helper(role => sub {
       my ($c, $extradata) = @_;
        return $user_role_cb->($c, $extradata);
    });
}
1;
__END__
=pod

=head1 NAME

Mojolicious::Plugin::Authorization - A plugin to make Authorization a bit easier

=head1 VERSION

version 1.05

=head1 SYNOPSIS

    use Mojolicious::Plugin::Authorization
    $self->plugin('Authorization' => {
        'has_priv'    => sub { ... },
        'is_role'     => sub { ... },
        'user_privs'  => sub { ... },
        'user_role'   => sub { ... },
        'fail_render' => { status => 401, json => { ... } },
    });
    if ($self->has_priv('delete_all', { optional => 'extra data stuff' })) {
        ...
    }

=head1 DESCRIPTION

A very simple API implementation of role-based access control (RBAC). This plugin is only an API you will
have to do all the work of setting up your roles and privileges and then provide four subs that are used by
the plugin.
The plugin expects that the current session will be used to get the role its privileges. It also assumes that
you have already been authenticated and your role set.
That is about it you are free to implement any system you like.

=head1 METHODS

=head2 has_priv('privilege', $extra_data) or has_privilege('privilege', $extra_data)

'has_priv' and 'has_privilege' will use the supplied C<has_priv> subroutine ref to check if the current session has the
given privilege. Returns true when the session has the privilege or false otherwise.
You can pass additional data along in the extra_data hashref and it will be passed to your C<has_priv>
subroutine as-is.

=head2 is('role',$extra_data) or is_role('role',$extra_data)

'is' and 'is_role' will use the supplied C<is_role> subroutine ref to check if the current session is the
given role. Returns true when the session has privilege or false otherwise.
You can pass additional data along in the extra_data hashref and it will be passed to your C<is_role>
subroutine as-is.

=head2 privileges($extra_data)

'privileges' will use the supplied C<user_privs> subroutine ref and return the privileges of the current session.
You can pass additional data along in the extra_data hashref and it will be passed to your C<user_privs>
subroutine as-is. The returned data is dependent on the supplied C<user_privs> subroutine.

=head2 role($extra_data)

'role' will use the supplied C<user_role> subroutine ref and return the role of the current session.
You can pass additional data along in the extra_data hashref and it will be passed to your C<user_role>
subroutine as-is. The returned data is dependent on the supplied C<user_role> subroutine.

=head1 CONFIGURATION

The following options must be set for the plugin:

=over 4

=item has_priv (REQUIRED) A coderef for checking to see if the current session has a privilege (see L</"HAS PRIV">).

=item is_role (REQUIRED) A coderef for checking to see if the current session is a certain role (see L</"IS / IS ROLE">).

=item user_privs (REQUIRED) A coderef for returning the privileges of the current session (see L</"PRIVILEGES">).

=item user_role (REQUIRED) A coderef for returning the role of the current session (see L</"ROLE">).

=back

The following options are not required but allow greater control:

=over 4

=item fail_render (OPTIONAL) A hashref for setting the status code and rendering json/text/etc when routing fails (see L</"ROUTING VIA CALLBACK">).

=back

=head2 HAS PRIV / HAS PRIVILEGE

'has_priv' is used when you need to confirm that the current session has the given privilege.
The coderef you pass to the C<has_priv> configuration key has the following signature:

    sub {
        my ($app, $privilege,$extradata) = @_;
        ...
    }

You must return either 0 for a fail and 1 for a pass.  This allows C<ROUTING VIA CONDITION> to work correctly.

=head2 IS / IS ROLE

'is' / 'is_role' is used when you need to confirm that the current session is set to the given role.
The coderef you pass to the C<is_role> configuration key has the following signature:

    sub {
        my ($app, $role, $extradata) = @_;
        ...
        return $role;
    }

You must return either 0 for a fail and 1 for a pass.  This allows C<ROUTING VIA CONDITION> to work correctly.

=head2 PRIVILEGES

'privileges' is used when you need to get all the privileges of the current session.
The coderef you pass to the C<user_privs> configuration key has the following signature:

    sub {
        my ($app,$extradata) = @_;
        ...
        return $privileges;
    }

You can return anything you want. It would normally be an arrayref of privileges but you are free to
return a scalar, hashref, arrayref, blessed object, or undef.

=head2 ROLE

'role' is used when you need to get the role of the current session.
The coderef you pass to the C<user_privs> configuration key has the following signature:

    sub {
        my ($app,$extradata) = @_;
        ...
        return $role;
    }

You can return anything you want. It would normally be just a scalar but you are free to
return a scalar, hashref, arrayref, blessed object, or undef.

=head1 EXAMPLES

For a code example using this, see the F<t/01-functional.t> test,
it uses L<Mojolicious::Lite> and this plugin.

=head1 ROUTING VIA CONDITION

This plugin also exports a routing condition you can use in order to limit access to certain documents to only
sessions that have a privilege.

    $r->route('/delete_all')->over(has_priv => 'delete_all')->to('mycontroller#delete_all');
    $r->route('/delete_all')->over(has_priv => ['delete_all', $extradata])->to('mycontroller#delete_all');
    my $delete_all_only = $r->route('/members')->over(has_priv => 'delete_all')->to('members#delete_all');
    $delete_all_only->route('delete')->to('members#delete_all');

If the session does not have the 'delete_all' privilege, these routes will not be considered by the dispatcher and unless you have set up a catch-all route,
 a 404 Not Found will be generated instead.

Another condition you can use to limit access to certain documents to only those sessions that
have a role.

    $r->route('/view_all')->over(is => 'ADMIN')->to('mycontroller#view_all');
    $r->route('/view_all')->over(is => ['ADMIN', $extradata])->to('mycontroller#view_all');
    my $view_all_only = $r->route('/members')->over(is => 'view_all')->to('members#view_all');
    $view_all_only->route('view')->to('members#view_all');

If the session is not the 'ADMIN' role, these routes will not be considered by the dispatcher and unless you have set up a catch-all route,
 a 404 Not Found will be generated instead.
This behavior is similar to the "has" condition.

=head1 ROUTING VIA CALLBACK

It is not recommended to route un-authorized requests to anything but a 404 page. If you do route to some sort
of 'You are not allowed page' you are telling a hacker that the URL was correct while the 404 tells them nothing.
This is just my opinion.

However in the case of publicly documented APIs returning a 404 when priv/role checks fails can confuse users, so
you can override the default 404 status on failure by supplying a 'fail_render' value in the plugin config. This
will be passed to the Mojolicious ->render method when the has_priv/is/is_role routing fails. For example, to
return a status code of 401 with JSON:

    fail_render => { status => 401, json => { error => 'Denied' } },

=head1 SEE ALSO

L<Mojolicious::Sessions>, L<Mojocast 3: Authorization|http://mojocasts.com/e3#>

=head1 AUTHOR

John Scoles, C<< <byterock  at hotmail.com> >>

=head1 BUGS / CONTRIBUTING

Please report any bugs or feature requests through the web interface at L<https://github.com/byterock/mojolicious-plugin-authorization/issues>.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.
    perldoc Mojolicious::Plugin::Authorization
You can also look for information at:

=over 4

=item * AnnoCPAN: Annotated CPAN documentation L<http://annocpan.org/dist/Mojolicious-Plugin-Authorization>

=item * CPAN Ratings L<http://cpanratings.perl.org/d/Mojolicious-Plugin-Authorization>

=item * Search CPAN L<http://search.cpan.org/dist/Mojolicious-Plugin-Authorization/>

=back

=head1 ACKNOWLEDGEMENTS

Ben van Staveren   (madcat)

    -   For 'Mojolicious::Plugin::Authentication' which I used as a guide in writing up this one.

Chuck Finley

    -   For staring me off on this.

Abhijit Menon-Sen

    -   For the routing suggestions

Roland Lammel

    -   For some other good suggestions

Lee Johnson

    -   For the latest updates for version 1.04
    
Matt Altus

    -   For the Allow extra data via conditional routing enhancement 
    
=head1 LICENSE AND COPYRIGHT

Copyright 2012 John Scoles.
This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.
See http://dev.perl.org/licenses/ for more information.

=cut
